### PR TITLE
[PM-17430] Implement new key format

### PR DESCRIPTION
**Pull request by quexten**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-17430

## 📔 Objective

Adds a new serialization format for symmetric keys. _**Read only in this PR**._

Historically, keys have been encoded directly as byte arrays, without describing their type. They can be categorized by size only; a 32 byte key is for encrypting AES256-CBC (No HMAC), a 64 byte key is for encrypting AES256-CBC-HMAC.

To distinguish these, the new key format now self-describes the key type. The struct contains the key type, and the key data, and is serialized into the final byte array using cbor. Cbor serialization is also used in e.g. passkeys.

Since it *could* be the case that a 32 byte cbor serialized key with type information ends up being 64 bytes, padding is introduced, ensuring the keys are always 65 or more bytes.

Specifically, the padding looks as follows:
```
N|0x00...0x00|KEY_BYTES
  ^ N bytes
```
where N is chosen such that the padded key is at least 65 bytes, ensuring that it cannot be confused with other keys.

The key itself looks as follows:
```
#[derive(Clone, Serialize, Deserialize)]
#[serde(tag = "alg")]
enum SerializedSymmetricCryptoKey {
    #[serde(rename = "A256C")]
    Aes256Cbc {
        #[serde(with = "serde_bytes", rename = "ek")]
        encryption_key: Vec<u8>,
    },
    #[serde(rename = "A256C-H")]
    Aes256CbcHmac {
        #[serde(with = "serde_bytes", rename = "ek")]
        encryption_key: Vec<u8>,
        #[serde(with = "serde_bytes", rename = "ak")]
        authentication_key: Vec<u8>,
    },
}

```

The considerations here are as follows:

https://bitwarden.atlassian.net/wiki/spaces/EN/pages/1557561420/Key+Crypto+Serialization+Format

- The encoding size matters, thus a format that serializes cryptographic ciphertexts as byte arrays well is chosen, rather than a format that needs text-based (double) encoding to b64. In this case we already use cbor, and it is a well-established format. It self-describes the structure in contrast to e.g. protobuf, which means there are no implicit assumptions in the format made by the encoding side, but explicitly the format is described.
for encoding cryptographic ciphertexts or keys
- The algorithm choices are abbreviated to save space during serialization
- The key (algorithm) field is shortened to "alg" in the serialization

The "algorithm" field allows adding as many encryption algorithms (or formats) as needed in the future, since it would just be another entry in the enum.

> [!NOTE]  
> This implements PartialEq for symmetric keys and makes it constant time. This is somewhat out of scope of this PR but was required to fix a section of code that otherwise could break later, when we introduce new keys that *only* use the new serialization format.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation
  team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: crates/bitwarden-crypto/src/keys/symmetric_crypto_key.rs [hunk 1]
```diff
@@ -53,31 +78,48 @@ impl SymmetricCryptoKey {
...
+            (SymmetricCryptoKey::Aes256CbcHmacKey(key), false) => {
+                let mut buf = Vec::with_capacity(64);
+                buf.extend_from_slice(&key.enc_key);
+                buf.extend_from_slice(&key.mac_key);
+                Ok(buf)
+            }
+            (_, true) => {
+                let serialized_key = SerializedSymmetricCryptoKey::from(self.clone());
+                let mut encoded_key = Vec::with_capacity(1024);
+                ciborium::into_writer(&serialized_key, &mut encoded_key)
```
Comment (dani-garcia) [inline crates/bitwarden-crypto/src/keys/symmetric_crypto_key.rs]: Two things:
I don't think we need to preallocate the Vec, it seems like ciborium already does it for you according to some tests: https://github.com/enarx/ciborium/blob/f3671472f80a67008f2f90ef33b185d8d81d703d/ciborium/tests/canonical.rs#L57-L58. If we decide to preallocate, we should probably choose a smaller size, otherwise we're wasting a fair amount of space.

What are the error conditions for `ciborium::into_writer`? Having to expose a result type if it can never happen seems very annoying, specially if it's something that can't ever happen in practice. At least with `serde_json::to_string`, it can only really fail in very specific circumstances (the only case I've seen were hashmaps with non-string keys, which are invalid JSON). If the errors can only appear when `SerializedSymmetricCryptoKey` has an invalid CBOR representation, I think we should just unwrap the error and add a test to make sure it's not broken in the future.

File: crates/bitwarden-crypto/src/keys/symmetric_crypto_key.rs [hunk 2]
```diff
@@ -141,6 +200,58 @@ impl std::fmt::Debug for SymmetricCryptoKey {
...
+            }
+        }
+    }
+}
+
+/// Pad a key to a minimum length;
+/// The first byte describes the number (N) of subsequently following null bytes
+/// Next, there are N null bytes
+/// Finally, the key bytes are appended
+fn pad_key(key_bytes: &[u8], min_length: usize) -> Vec<u8> {
```
Comment (dani-garcia) [inline crates/bitwarden-crypto/src/keys/symmetric_crypto_key.rs]: Does the padding need to be at the start? If we put it at the end we can modify the vector in place and avoid making new allocations.

File: crates/bitwarden-crypto/src/keys/symmetric_crypto_key.rs [hunk 3]
```diff
@@ -106,23 +148,40 @@ impl TryFrom<&mut [u8]> for SymmetricCryptoKey {
...
 
             Ok(SymmetricCryptoKey::Aes256CbcKey(Aes256CbcKey { enc_key }))
+        } else if value.len() > Self::AES256_CBC_HMAC_KEY_LEN {
+            let unpadded_value = zeroize::Zeroizing::new(unpad_key(value));
+            let decoded_key: SerializedSymmetricCryptoKey =
+                ciborium::from_reader(unpadded_value.as_slice())
+                    .map_err(|_| CryptoError::EncodingError)?;
+            value.zeroize();
+
+            match decoded_key {
```
Comment (dani-garcia) [inline crates/bitwarden-crypto/src/keys/symmetric_crypto_key.rs]: This should probably be moved into an `impl TryFrom<SerializedSymmetricCryptoKey> for SymmetricCryptoKey`

File: crates/bitwarden-crypto/src/keys/symmetric_crypto_key.rs [hunk 4]
```diff
@@ -141,6 +200,58 @@ impl std::fmt::Debug for SymmetricCryptoKey {
...
+    let null_bytes = min_length.saturating_sub(1).saturating_sub(key_bytes.len());
+    let mut padded_key = Vec::with_capacity(min_length);
+    padded_key.extend_from_slice(&[null_bytes as u8]);
+    padded_key.extend_from_slice(vec![0u8; null_bytes].as_slice());
+    padded_key.extend_from_slice(key_bytes);
+    padded_key
+}
+
+// Unpad a key that was padded with pad_key
+fn unpad_key(key_bytes: &[u8]) -> Vec<u8> {
```
Comment (dani-garcia) [inline crates/bitwarden-crypto/src/keys/symmetric_crypto_key.rs]: We could change the interface here to be `fn unpad_key(key_bytes: &[u8]) -> &[u8]` and avoid an extra allocation as well. This would also save us from having to wrap it on Zeroizing when using it.

File: crates/bitwarden-crypto/src/keys/symmetric_crypto_key.rs (hunk 2 above)
Comment (quexten) [reply crates/bitwarden-crypto/src/keys/symmetric_crypto_key.rs]: That is fine. So do I understand correctly that the requested format would be:
```
payload|0x00...0x00|pad_len
             ^ pad_len times 0x00 bytes
```

Comment (dani-garcia) [reply crates/bitwarden-crypto/src/keys/symmetric_crypto_key.rs]: Yeah, that was my thinking. We can also do like `PKCS7` padding and set the pad values to `pad_len`, which simplifies the implementation a bit:
```rust
fn pad_key(key_bytes: &mut Vec<u8>, min_length: usize) {
    let pad_bytes = min_length.saturating_sub(key_bytes.len());
    if pad_bytes > 0 {
        key_bytes.resize(min_length, pad_bytes as u8);
    }
}
```

Comment (quexten) [reply crates/bitwarden-crypto/src/keys/symmetric_crypto_key.rs]: This didn't seem to work quite right. Padding should happen in all cases, for at least one byte, otherwise we cannot reliably tell whether something was padded or not.

```
/// Pad a key to a minimum length using PKCS7-like padding
fn pad_key(key_bytes: &mut Vec<u8>, min_length: usize) {
    // at least 1 byte of padding is required
    let pad_bytes = min_length.saturating_sub(key_bytes.len()).max(1);
    let padded_length = max(min_length, key_bytes.len() + 1);
    key_bytes.resize(padded_length, pad_bytes as u8);
}
```
seems to work though, with added tests for the boundary values.

File: crates/bitwarden-crypto/src/keys/symmetric_crypto_key.rs (hunk 3 above)
Comment (quexten) [reply crates/bitwarden-crypto/src/keys/symmetric_crypto_key.rs]: Question: If we use `TryFrom` we expect error cases. What would be the error cases in this case? I can imagine `GenericArray::from_slice` failing because for whatever reason the `Vec` from the SerializedCryptoKey has the wrong size? Is there anything else?

File: crates/bitwarden-crypto/src/keys/symmetric_crypto_key.rs (hunk 1 above)
Comment (quexten) [reply crates/bitwarden-crypto/src/keys/symmetric_crypto_key.rs]: The two conditions I found are: https://github.com/enarx/ciborium/blob/f3671472f80a67008f2f90ef33b185d8d81d703d/ciborium/src/ser/mod.rs#L417 and if the underlying writer errors (which shouldn't happen for a vec?).
So, unwrap seems fine to me.

File: crates/bitwarden-crypto/src/keys/symmetric_crypto_key.rs (hunk 3 above)
Comment (dani-garcia) [reply crates/bitwarden-crypto/src/keys/symmetric_crypto_key.rs]: I was thinking for the `_ => Err(CryptoError::InvalidKey),` variant, but we're probably better off implementing `From` and just implementing the missing `Aes256Cbc` variant?

File: crates/bitwarden-crypto/src/keys/symmetric_crypto_key.rs (hunk 1 above)
Comment (quexten) [reply crates/bitwarden-crypto/src/keys/symmetric_crypto_key.rs]: (Updated to unwrap and not use `Result`)

File: crates/bitwarden-crypto/src/keys/symmetric_crypto_key.rs (hunk 2 above)
Comment (dani-garcia) [reply crates/bitwarden-crypto/src/keys/symmetric_crypto_key.rs]: Oh yep, I forgot that we always want to pad at least once, your example looks good to me.
